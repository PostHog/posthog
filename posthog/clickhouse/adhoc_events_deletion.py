from posthog.clickhouse.cluster import ON_CLUSTER_CLAUSE
from posthog.clickhouse.table_engines import (
    ReplacingMergeTreeDeleted,
)


ADHOC_EVENTS_DELETION_TABLE = "adhoc_events_deletion"


def ADHOC_EVENTS_DELETION_TABLE_SQL(on_cluster=True):
    return """
CREATE TABLE IF NOT EXISTS {table_name} {on_cluster_clause}
(
    team_id Int64,
    uuid UUID,
    created_at DateTime64(6, 'UTC') DEFAULT now64(),
    deleted_at DateTime,
    is_deleted UInt8 DEFAULT 0
) ENGINE = {engine}
order by (team_id, uuid)
TTL deleted_at + INTERVAL 3 MONTH WHERE is_deleted = 1
""".format(
        table_name=ADHOC_EVENTS_DELETION_TABLE,
        on_cluster_clause=ON_CLUSTER_CLAUSE(on_cluster),
        engine=ReplacingMergeTreeDeleted(ADHOC_EVENTS_DELETION_TABLE, ver="deleted_at", is_deleted="is_deleted"),
    )


def DROP_ADHOC_EVENTS_DELETION_TABLE_SQL(on_cluster=True):
    return """
DROP TABLE IF EXISTS {table_name} {on_cluster_clause}
""".format(
        table_name=ADHOC_EVENTS_DELETION_TABLE,
        on_cluster_clause=ON_CLUSTER_CLAUSE(on_cluster),
    )
