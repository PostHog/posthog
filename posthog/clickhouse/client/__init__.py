from posthog.clickhouse.client.execute import query_with_columns, sync_execute
from posthog.clickhouse.client.execute_async import execute_with_progress

__all__ = [
    "sync_execute",
    "query_with_columns",
    "execute_with_progress",
]
