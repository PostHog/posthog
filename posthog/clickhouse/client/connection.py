from contextlib import contextmanager
from enum import Enum
from functools import lru_cache

from clickhouse_driver import Client as SyncClient
from clickhouse_pool import ChPool
from django.conf import settings


class Workload(Enum):
    # Default workload
    DEFAULT = "DEFAULT"
    # Analytics queries, other 'lively' queries
    ONLINE = "ONLINE"
    # Historical exports, other long-running processes where latency is less critical
    OFFLINE = "OFFLINE"


_default_workload = Workload.ONLINE


def get_pool(workload: Workload, team_id=None, readonly=False):
    """
    Returns the right connection pool given a workload.

    Note that the same pool should be returned every call.
    """
    if team_id is not None and str(team_id) in settings.CLICKHOUSE_PER_TEAM_SETTINGS:
        return make_ch_pool(**settings.CLICKHOUSE_PER_TEAM_SETTINGS[str(team_id)])

    # Note that `readonly` does nothing if the relevant vars are not set!
    if readonly and settings.READONLY_CLICKHOUSE_USER is not None and settings.READONLY_CLICKHOUSE_PASSWORD:
        return make_ch_pool(
            user=settings.READONLY_CLICKHOUSE_USER,
            password=settings.READONLY_CLICKHOUSE_PASSWORD,
        )

    if (
        workload == Workload.OFFLINE or workload == Workload.DEFAULT and _default_workload == Workload.OFFLINE
    ) and settings.CLICKHOUSE_OFFLINE_CLUSTER_HOST is not None:
        return make_ch_pool(host=settings.CLICKHOUSE_OFFLINE_CLUSTER_HOST)

    return make_ch_pool()


def default_client():
    """
    Return a bare bones client for use in places where we are only interested in general ClickHouse state
    DO NOT USE THIS FOR QUERYING DATA
    """
    return SyncClient(
        host=settings.CLICKHOUSE_HOST,
        # We set "system" here as we don't necessarily have a "default" database,
        # which is what the clickhouse_driver would use by default. We are
        # assuming that this exists and we have permissions to access it. This
        # feels like a reasonably safe assumption as e.g. we already reference
        # `system.numbers` in multiple places within queries. We also assume
        # access to various other tables e.g. to handle async migrations.
        database="system",
        secure=settings.CLICKHOUSE_SECURE,
        user=settings.CLICKHOUSE_USER,
        password=settings.CLICKHOUSE_PASSWORD,
        ca_certs=settings.CLICKHOUSE_CA,
        verify=settings.CLICKHOUSE_VERIFY,
    )


@lru_cache(maxsize=None)
def make_ch_pool(**overrides) -> ChPool:
    kwargs = {
        "host": settings.CLICKHOUSE_HOST,
        "database": settings.CLICKHOUSE_DATABASE,
        "secure": settings.CLICKHOUSE_SECURE,
        "user": settings.CLICKHOUSE_USER,
        "password": settings.CLICKHOUSE_PASSWORD,
        "ca_certs": settings.CLICKHOUSE_CA,
        "verify": settings.CLICKHOUSE_VERIFY,
        "connections_min": settings.CLICKHOUSE_CONN_POOL_MIN,
        "connections_max": settings.CLICKHOUSE_CONN_POOL_MAX,
        "settings": {"mutations_sync": "1"} if settings.TEST else {},
        # Without this, OPTIMIZE table and other queries will regularly run into timeouts
        "send_receive_timeout": 30 if settings.TEST else 999_999_999,
        **overrides,
    }

    return ChPool(**kwargs)


@contextmanager
def set_default_clickhouse_workload_type(workload: Workload):
    global _default_workload

    _default_workload = workload


ch_pool = get_pool(workload=Workload.ONLINE)
