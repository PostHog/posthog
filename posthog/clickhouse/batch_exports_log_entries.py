from posthog.clickhouse.kafka_engine import KAFKA_COLUMNS, kafka_engine, ttl_period
from posthog.clickhouse.table_engines import ReplacingMergeTree
from posthog.kafka_client.topics import KAFKA_BATCH_EXPORTS_LOG_ENTRIES
from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_DATABASE

BATCH_EXPORTS_LOG_ENTRIES_TABLE = "batch_exports_log_entries"
BATCH_EXPORTS_LOG_ENTRIES_TTL_WEEKS = 1

BATCH_EXPORTS_LOG_ENTRIES_TABLE_BASE_SQL = """
CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
(
    team_id UInt64,
    batch_export_id VARCHAR,
    run_id VARCHAR,
    timestamp DateTime64(6, 'UTC'),
    level VARCHAR,
    message VARCHAR
    {extra_fields}
) ENGINE = {engine}
"""

BATCH_EXPORTS_LOG_ENTRIES_TABLE_ENGINE = lambda: ReplacingMergeTree(BATCH_EXPORTS_LOG_ENTRIES_TABLE, ver="_timestamp")
BATCH_EXPORTS_LOG_ENTRIES_TABLE_SQL = lambda: (
    BATCH_EXPORTS_LOG_ENTRIES_TABLE_BASE_SQL
    + """ORDER BY (team_id, batch_export_id, run_id, timestamp)
{ttl_period}
SETTINGS index_granularity=512
"""
).format(
    table_name=BATCH_EXPORTS_LOG_ENTRIES_TABLE,
    cluster=CLICKHOUSE_CLUSTER,
    extra_fields=KAFKA_COLUMNS,
    engine=BATCH_EXPORTS_LOG_ENTRIES_TABLE_ENGINE(),
    ttl_period=ttl_period("timestamp", BATCH_EXPORTS_LOG_ENTRIES_TTL_WEEKS),
)

KAFKA_BATCH_EXPORTS_LOG_ENTRIES_TABLE_SQL = lambda: BATCH_EXPORTS_LOG_ENTRIES_TABLE_BASE_SQL.format(
    table_name="kafka_" + BATCH_EXPORTS_LOG_ENTRIES_TABLE,
    cluster=CLICKHOUSE_CLUSTER,
    engine=kafka_engine(topic=KAFKA_BATCH_EXPORTS_LOG_ENTRIES),
    extra_fields="",
)

BATCH_EXPORTS_LOG_ENTRIES_TABLE_MV_SQL = """
CREATE MATERIALIZED VIEW IF NOT EXISTS {table_name}_mv ON CLUSTER '{cluster}'
TO {database}.{table_name}
AS SELECT
team_id,
batch_export_id,
run_id,
timestamp,
level,
message,
_timestamp,
_offset
FROM {database}.kafka_{table_name}
""".format(
    table_name=BATCH_EXPORTS_LOG_ENTRIES_TABLE, cluster=CLICKHOUSE_CLUSTER, database=CLICKHOUSE_DATABASE
)


INSERT_BATCH_EXPORTS_LOG_ENTRY_SQL = """
INSERT INTO batch_exports_log_entries SELECT %(team_id)s, %(batch_export_id)s, %(run_id)s, %(timestamp)s, %(level)s, %(message)s, now(), 0
"""

TRUNCATE_BATCH_EXPORTS_LOG_ENTRIES_TABLE_SQL = (
    f"TRUNCATE TABLE IF EXISTS {BATCH_EXPORTS_LOG_ENTRIES_TABLE} ON CLUSTER '{CLICKHOUSE_CLUSTER}'"
)
