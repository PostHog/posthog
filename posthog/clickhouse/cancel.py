from statshog.defaults.django import statsd

from posthog.api.services.query import logger
from posthog.clickhouse.client import sync_execute
from posthog.settings import CLICKHOUSE_CLUSTER


def cancel_query_on_cluster(team_id: int, client_query_id: str) -> None:
    result = sync_execute(
        f"KILL QUERY ON CLUSTER '{CLICKHOUSE_CLUSTER}' WHERE query_id LIKE %(client_query_id)s",
        {"client_query_id": f"{team_id}_{client_query_id}%"},
    )
    logger.info("Cancelled query %s for team %s, result: %s", client_query_id, team_id, result)
    statsd.incr("clickhouse.query.cancellation_requested", tags={"team_id": team_id})
