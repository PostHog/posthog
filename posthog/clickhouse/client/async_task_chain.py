import threading
import typing
from contextlib import contextmanager
from functools import partial

from celery import chain
from celery.canvas import Signature
from django.db import transaction

from posthog.schema import QueryStatus

if typing.TYPE_CHECKING:
    from posthog.clickhouse.client.execute_async import QueryStatusManager

# Create a thread-local storage for task chains
_thread_locals = threading.local()


def kick_off_task(
    manager: "QueryStatusManager",
    query_status: QueryStatus,
    task_signature: Signature,
) -> None:
    task = task_signature.apply_async()
    query_status.task_id = task.id
    manager.store_query_status(query_status)


def get_task_chain() -> list[tuple[Signature, "QueryStatusManager", QueryStatus]]:
    """
    Retrieves the task chain from thread-local storage.
    """
    if not hasattr(_thread_locals, "task_chain"):
        _thread_locals.task_chain = []
    return _thread_locals.task_chain


def set_in_context(value) -> None:
    """
    Sets the in_context flag in thread-local storage.
    """
    _thread_locals.in_context = value


def is_in_context() -> bool:
    """
    Checks if the in_context flag is set in thread-local storage.
    """
    return getattr(_thread_locals, "in_context", False)


def add_task_to_chain(task_signature: Signature, manager: "QueryStatusManager", query_status: QueryStatus) -> None:
    """
    Adds a task to the chain. If not in context, registers the task with on_commit directly.
    """

    if is_in_context():
        task_chain = get_task_chain()
        task_chain.append((task_signature, manager, query_status))
    else:
        transaction.on_commit(partial(kick_off_task, manager, query_status, task_signature))


def execute_task_chain() -> None:
    """
    Executes the task chain after the transaction is committed.
    """
    task_chain = get_task_chain()
    if task_chain:
        chained_tasks = chain(*[args[0] for args in task_chain])
        result = chained_tasks.apply_async()

        for args in task_chain:
            args[2].task_id = result.id
            args[1].store_query_status(args[2])

        _thread_locals.task_chain = []


@contextmanager
def task_chain_context() -> typing.Iterator[None]:
    set_in_context(True)
    try:
        yield
    finally:
        set_in_context(False)
        transaction.on_commit(execute_task_chain)
