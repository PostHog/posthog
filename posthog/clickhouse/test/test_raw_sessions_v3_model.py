import datetime

from posthog.test.base import (
    BaseTest,
    ClickhouseTestMixin,
    _create_event,
    flush_persons_and_events,
    snapshot_clickhouse_queries,
)

from posthog.clickhouse.client import query_with_columns, sync_execute
from posthog.models.raw_sessions.sql_v3 import RAW_SESSION_TABLE_BACKFILL_SQL_V3
from posthog.models.utils import uuid7

distinct_id_counter = 0
session_id_counter = 0


def create_distinct_id():
    global distinct_id_counter
    distinct_id_counter += 1
    return f"d{distinct_id_counter}"


def create_session_id():
    global session_id_counter
    session_id_counter += 1
    return str(uuid7(random=session_id_counter))


@snapshot_clickhouse_queries
class TestRawSessionsModel(ClickhouseTestMixin, BaseTest):
    snapshot_replace_all_numbers = True

    def select_by_session_id(self, session_id):
        flush_persons_and_events()
        return query_with_columns(
            """
            select
                *
            from raw_sessions_v3_v
            where
                session_id_v7 = toUInt128(toUUID(%(session_id)s)) AND
                team_id = %(team_id)s
                """,
            {
                "session_id": session_id,
                "team_id": self.team.id,
            },
        )

    def test_it_creates_session_when_creating_event(self):
        distinct_id = create_distinct_id()
        session_id = create_session_id()
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$current_url": "/", "$session_id": session_id},
            timestamp="2024-03-08",
        )

        response = sync_execute(
            """
            select
                session_id_v7,
                team_id
            from raw_sessions_v3_v
            where
                session_id_v7 = toUInt128(toUUID(%(session_id)s))  AND
                team_id = %(team_id)s
                """,
            {
                "session_id": session_id,
                "team_id": self.team.id,
            },
        )

        assert len(response) == 1

    def test_handles_different_distinct_id_across_same_session(self):
        distinct_id1 = create_distinct_id()
        distinct_id2 = create_distinct_id()
        session_id = create_session_id()

        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id1,
            properties={"$session_id": session_id},
            timestamp="2024-03-08",
        )
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id2,
            properties={"$session_id": session_id},
            timestamp="2024-03-08",
        )

        responses = self.select_by_session_id(session_id)
        assert len(responses) == 1
        assert responses[0]["distinct_id"] in {distinct_id1, distinct_id2}
        assert responses[0]["pageview_uniq"] == 2

    def test_handles_entry_and_exit_urls(self):
        distinct_id = create_distinct_id()
        session_id = create_session_id()

        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$current_url": "/entry", "$session_id": session_id},
            timestamp="2024-03-08:01",
        )
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$current_url": "/middle", "$session_id": session_id},
            timestamp="2024-03-08:02",
        )
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$current_url": "/middle", "$session_id": session_id},
            timestamp="2024-03-08:03",
        )
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$current_url": "/exit", "$session_id": session_id},
            timestamp="2024-03-08:04",
        )

        responses = self.select_by_session_id(session_id)
        assert len(responses) == 1
        assert responses[0]["entry_url"] == "/entry"
        assert responses[0]["end_url"] == "/exit"
        assert len(responses[0]["urls"]) == 3
        assert set(responses[0]["urls"]) == {"/entry", "/middle", "/exit"}  # order is not guaranteed

    def test_handles_initial_utm_properties(self):
        distinct_id = create_distinct_id()
        session_id = create_session_id()

        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$session_id": session_id, "utm_source": "source"},
            timestamp="2024-03-08",
        )
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$session_id": session_id, "utm_source": "other_source"},
            timestamp="2024-03-08",
        )

        responses = self.select_by_session_id(session_id)
        assert len(responses) == 1
        assert responses[0]["entry_utm_source"] == "source"

    def test_counts_pageviews_autocaptures_and_events(self):
        distinct_id = create_distinct_id()
        session_id = create_session_id()

        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$session_id": session_id},
            timestamp="2024-03-08",
        )
        _create_event(
            team=self.team,
            event="$autocapture",
            distinct_id=distinct_id,
            properties={"$session_id": session_id},
            timestamp="2024-03-08",
        )
        _create_event(
            team=self.team,
            event="$autocapture",
            distinct_id=distinct_id,
            properties={"$session_id": session_id},
            timestamp="2024-03-08",
        )
        _create_event(
            team=self.team,
            event="other event",
            distinct_id=distinct_id,
            properties={"$session_id": session_id},
            timestamp="2024-03-08",
        )
        _create_event(
            team=self.team,
            event="$pageleave",
            distinct_id=distinct_id,
            properties={"$session_id": session_id},
            timestamp="2024-03-08",
        )

        responses = self.select_by_session_id(session_id)
        assert len(responses) == 1
        assert responses[0]["pageview_uniq"] == 1
        assert responses[0]["autocapture_uniq"] == 2

    def test_separates_sessions_across_same_user(self):
        distinct_id = create_distinct_id()
        session_id1 = create_session_id()
        session_id2 = create_session_id()
        session_id3 = create_session_id()

        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$session_id": session_id1},
            timestamp="2024-03-08",
        )
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$session_id": session_id2},
            timestamp="2024-03-08",
        )

        responses = self.select_by_session_id(session_id1)
        assert len(responses) == 1
        responses = self.select_by_session_id(session_id2)
        assert len(responses) == 1
        responses = self.select_by_session_id(session_id3)
        assert len(responses) == 0

    def test_select_from_sessions(self):
        # just make sure that we can select from the sessions table without error
        distinct_id = create_distinct_id()
        session_id = create_session_id()
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$session_id": session_id},
            timestamp="2024-03-08",
        )

        # we can't include all the columns as this clickhouse driver doesn't support selecting states
        responses = sync_execute(
            """
        SELECT
            session_id_v7,
            team_id,
            min_timestamp,
            max_timestamp,
            urls
        FROM raw_sessions_v3
        WHERE session_id_v7 = toUInt128(toUUID(%(session_id)s)) AND team_id = %(team_id)s
        """,
            {
                "session_id": session_id,
                "team_id": self.team.id,
            },
        )
        assert len(responses) == 1

    def test_select_from_sessions_mv(self):
        # just make sure that we can select from the sessions mv without error
        distinct_id = create_distinct_id()
        session_id = create_session_id()
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$session_id": session_id},
            timestamp="2024-03-08",
        )

        # we can't include all the columns as this clickhouse driver doesn't support selecting states
        responses = sync_execute(
            """
        SELECT
            session_id_v7,
            team_id,
            min_timestamp,
            max_timestamp,
            urls
        FROM raw_sessions_v3_mv
        WHERE session_id_v7 = toUInt128(toUUID(%(session_id)s)) AND team_id = %(team_id)s
        """,
            {
                "session_id": session_id,
                "team_id": self.team.id,
            },
        )
        assert len(responses) == 1

    def test_backfill_sql(self):
        distinct_id = create_distinct_id()
        session_id = create_session_id()
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$current_url": "/", "$session_id": session_id},
            timestamp="2024-03-08",
        )

        # just test that the backfill SQL can be run without error
        sync_execute(
            RAW_SESSION_TABLE_BACKFILL_SQL_V3("team_id = %(team_id)s"),
            {"team_id": self.team.id},
        )

    def test_max_inserted_at(self):
        distinct_id = create_distinct_id()
        session_id = create_session_id()
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$session_id": session_id},
            timestamp="2024-03-08",
        )
        now = datetime.datetime.now(datetime.UTC)

        result = self.select_by_session_id(session_id)
        max_inserted_at = result[0]["max_inserted_at"]
        # assert that it's close to now, allowing for a small margin of error because we're running this on CI in the cloud somewhere with preempting
        assert abs((max_inserted_at - now).total_seconds()) < 10

    def test_ad_ids_map_and_set(self):
        distinct_id = create_distinct_id()
        session_id = create_session_id()

        present_ad_id = "irclid"
        missing_ad_id = "wbraid"
        value = "test_irclid"

        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={"$session_id": session_id, present_ad_id: value},
            timestamp="2024-03-08",
        )

        result = self.select_by_session_id(session_id)

        assert result[0]["entry_ad_ids_map"][present_ad_id] == value
        assert missing_ad_id not in result[0]["entry_ad_ids_map"].keys()

        assert present_ad_id in result[0]["entry_ad_ids_set"]
        assert missing_ad_id not in result[0]["entry_ad_ids_set"]

    def test_channel_type_properties(self):
        distinct_id = create_distinct_id()
        session_id = create_session_id()

        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={
                "$session_id": session_id,
                "gad_source": "1",
                "gclid": "some_gclid",
                "$referring_domain": "google.com",
                "utm_campaign": "some_campaign",
                "utm_medium": "some_medium",
                "utm_source": "some_source",
            },
            timestamp="2024-03-08",
        )

        result = self.select_by_session_id(session_id)

        assert result[0]["entry_channel_type_properties"] == (
            "some_source",
            "some_medium",
            "some_campaign",
            "google.com",
            True,
            False,
            "1",
        )

    def test_autocapture_does_not_set_attribution_when_pageview_present(self):
        distinct_id = create_distinct_id()
        session_id = create_session_id()

        _create_event(
            team=self.team,
            event="$autocapture",
            distinct_id=distinct_id,
            properties={
                "$session_id": session_id,
                "$current_url": "/1",
                "utm_source": "source1",
            },
            timestamp="2024-03-08",
        )
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={
                "$session_id": session_id,
                "$current_url": "/2",
                "utm_source": "source2",
            },
            timestamp="2024-03-09",
        )

        result = self.select_by_session_id(session_id)

        assert result[0]["entry_url"] == "/2"
        assert result[0]["end_url"] == "/2"
        assert result[0]["urls"] == ["/2"]
        assert result[0]["entry_utm_source"] == "source2"

    def test_autocapture_does_set_attribution_when_only_event(self):
        distinct_id = create_distinct_id()
        session_id = create_session_id()

        _create_event(
            team=self.team,
            event="$autocapture",
            distinct_id=distinct_id,
            properties={
                "$session_id": session_id,
                "$current_url": "/1",
                "utm_source": "source1",
            },
            timestamp="2024-03-08",
        )

        result = self.select_by_session_id(session_id)

        assert result[0]["entry_url"] == "/1"
        assert result[0]["end_url"] == "/1"
        assert result[0]["urls"] == []
        assert result[0]["entry_utm_source"] == "source1"

    def test_store_all_feature_flag_values(self):
        distinct_id = create_distinct_id()
        session_id = create_session_id()

        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id,
            properties={
                "$session_id": session_id,
                "$feature/flag_string": "f1_a",
                "$feature/flag_int": 1,
                "$feature/flag_complex": ["hello", 123],
                "$feature/flag_duplicates": "a",
            },
            timestamp="2024-03-08",
        )

        _create_event(
            team=self.team,
            event="$autocapture",
            distinct_id=distinct_id,
            properties={
                "$session_id": session_id,
                "$feature/flag_string": "f1_b",
                "$feature/flag_int": 2,
                "$feature/flag_complex": {"key": "value"},
                "$feature/flag_duplicates": "a",
            },
            timestamp="2024-03-08",
        )

        result = self.select_by_session_id(session_id)

        # contains all values
        assert set(result[0]["flag_values"]["$feature/flag_string"]) == {"f1_a", "f1_b"}
        # converts to string
        assert set(result[0]["flag_values"]["$feature/flag_int"]) == {"1", "2"}
        # converts to json string
        assert set(result[0]["flag_values"]["$feature/flag_complex"]) == {'["hello",123]', '{"key":"value"}'}
        # deduplicates
        assert result[0]["flag_values"]["$feature/flag_duplicates"] == ["a"]

    def test_lookup_feature_flag(self):
        distinct_id_1 = create_distinct_id()
        session_id_1 = create_session_id()
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id_1,
            properties={
                "$session_id": session_id_1,
                "$feature/flag_string": "f1_a",
            },
            timestamp="2024-03-08",
        )
        _create_event(
            team=self.team,
            event="custom event",
            distinct_id=distinct_id_1,
            properties={
                "$session_id": session_id_1,
                "$feature/flag_string": "f1_b",
            },
            timestamp="2024-03-08",
        )

        distinct_id_2 = create_distinct_id()
        session_id_2 = create_session_id()
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id_2,
            properties={
                "$session_id": session_id_2,
                "$feature/flag_string": "f1_c",
            },
            timestamp="2024-03-08",
        )

        result = query_with_columns(
            """
            select
                session_id_v7,
                has(flag_values['$feature/flag_string'], 'f1_a') as has_f1_a,
                has(flag_values['$feature/flag_string'], 'f1_b') as has_f1_b,
                has(flag_values['$feature/flag_string'], 'f1_c') as has_f1_c
            from raw_sessions_v3_v
            where
                team_id = %(team_id)s
            ORDER BY session_id_v7
                """,
            {
                "team_id": self.team.id,
            },
        )
        assert result[0]["has_f1_a"]
        assert result[0]["has_f1_b"]
        assert not result[0]["has_f1_c"]

        assert not result[1]["has_f1_a"]
        assert not result[1]["has_f1_b"]
        assert result[1]["has_f1_c"]

    def test_tracks_all_distinct_ids(self):
        distinct_id_1 = create_distinct_id()
        distinct_id_2 = create_distinct_id()
        session_id = create_session_id()
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id_1,
            properties={"$current_url": "/", "$session_id": session_id},
            timestamp="2024-03-08",
        )
        _create_event(
            team=self.team,
            event="$pageview",
            distinct_id=distinct_id_2,
            properties={"$current_url": "/", "$session_id": session_id},
            timestamp="2024-03-08",
        )

        result = self.select_by_session_id(session_id)

        assert set(result[0]["distinct_ids"]) == {distinct_id_1, distinct_id_2}
