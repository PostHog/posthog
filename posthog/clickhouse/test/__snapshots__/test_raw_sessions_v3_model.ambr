# serializer version: 1
# name: TestRawSessionsModel.test_ad_ids_map_and_set
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_autocapture_does_not_set_attribution_when_pageview_present
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_autocapture_does_set_attribution_when_only_event
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_channel_type_properties
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_counts_pageviews_autocaptures_and_events
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_handles_different_distinct_id_across_same_session
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_handles_entry_and_exit_urls
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_handles_initial_utm_properties
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_has_replay_events
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_has_replay_events.1
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_it_creates_session_when_creating_event
  '''
  
  select session_id_v7,
         team_id
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_lookup_feature_flag
  '''
  
  select session_id_v7,
         has(flag_values['$feature/flag_string'], 'f1_a') as has_f1_a,
         has(flag_values['$feature/flag_string'], 'f1_b') as has_f1_b,
         has(flag_values['$feature/flag_string'], 'f1_c') as has_f1_c
  from raw_sessions_v3_v
  where team_id = 99999
  ORDER BY session_id_v7
  '''
# ---
# name: TestRawSessionsModel.test_max_inserted_at
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_select_from_sessions
  '''
  
  SELECT session_id_v7,
         team_id,
         min_timestamp,
         max_timestamp,
         urls
  FROM raw_sessions_v3
  WHERE session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_select_from_sessions_mv
  '''
  
  SELECT session_id_v7,
         team_id,
         min_timestamp,
         max_timestamp,
         urls
  FROM raw_sessions_v3_mv
  WHERE session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_separates_sessions_across_same_user
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_separates_sessions_across_same_user.1
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_separates_sessions_across_same_user.2
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_store_all_feature_flag_values
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
# name: TestRawSessionsModel.test_tracks_all_distinct_ids
  '''
  
  select *
  from raw_sessions_v3_v
  where session_id_v7 = toUInt128(toUUID('00000000-0000-0000-0000-000000000000'))
    AND team_id = 99999
  '''
# ---
